src/FunctionApps/DSTP/src/main/kotlin/gov/cdc/prime/phdi/Utilities/HelperFunctions.kt

```diff
@@ -0,0 +1,363 @@
+    val rootTemplate: String = when (messageType.lowercase()) {
+        "adt_a01" -> "ADT_A01"
+        "oml_o21" -> "OML_O21"
+        "oru_r01" -> "ORU_R01"
+        "vxu_v04" -> "VXU_V04"
+        "ccd" -> "CCD"
+        "ccda" -> "CCD"
+        "consultationnote" -> "ConsultationNote"
+        "dischargesummary" -> "DischargeSummary"
+        "historyandphysical" -> "HistoryandPhysical"
+        "operativenote" -> "OperativeNote"
+        "procedurenote" -> "ProcedureNote"
+        "progressnote" -> "ProgressNote"
+        "referralnote" -> "ReferralNote"
+        "transfersummary" -> "TransferSummary"
+        "examplepatient" -> "ExamplePatient"
+        "stu3chargeitem" -> "Stu3ChargeItem"
+        else -> "UNKNOWN"
+    }
+
+    val templateCollection: String = when (inputDataType) {
+        "Hl7v2" -> "microsofthealth/fhirconverter:default"
+        "Ccda" -> "microsofthealth/ccdatemplates:default"
+        "Json" -> "microsofthealth/jsontemplates:default"
+        else -> "microsofthealth/hl7v2templates:default"
+    }
+
+    if (inputDataType === "UNKNOWN" || rootTemplate === "UNKNOWN") {
+        return null
+    }
+
+    // connect to the FHIR $convert-data endpoint using the access token
+    val url: String = "${System.getenv("fhir_url")}/\$convert-data"
+    // use JSONObject instead of HashMap for seamless conversion to JSON-compliant strings
+    val requestBody: JSONObject = JSONObject()
+    requestBody.put("resourceType", "Parameters")
+    requestBody.put("parameter", listOf(
+        mapOf("name" to "inputData", "valueString" to "${message}"),
+        mapOf("name" to "inputDataType", "valueString" to "${inputDataType}"),
+        mapOf("name" to "templateCollectionReference", "valueString" to "${templateCollection}"),
+        mapOf("name" to "rootTemplate", "valueString" to "${rootTemplate}")
+    ))
+    val json: String = requestBody.toString()
+    
+    val client: HttpClient = HttpClient.newHttpClient()
+    val request: HttpRequest = HttpRequest.newBuilder()
+        .uri(URI.create(url))
+        .headers("Content-Type", "application/json", "Authorization", "Bearer ${accessToken}")
+        .POST(HttpRequest.BodyPublishers.ofString(json))
+        .build()
+    val response: HttpResponse<String> = client.send(
+        request,
+        HttpResponse.BodyHandlers.ofString()
+    )
+    
+    return response.body()
+}
+
+/* 
+   ********************** 
+    VALIDATING FUNCTIONS
+   **********************
+*/
+
+private fun isValidMessage(message: String, format: String): Boolean {
+    val context = DefaultHapiContext()
+    // We will likely replace this with a more custom validation class in the future
+    context.setValidationContext(ValidationContextFactory.defaultValidation() as ValidationContext)
+
+    val parser = when (format.lowercase()) {
+        "xml" -> context.getXMLParser()
+        "hl7" -> context.getPipeParser()
+        else -> context.getGenericParser()
+    }
```

Comment:
are we potentially using the other options?

Reply:
We're potentially using the `getXMLParser()` if we can overcome the issues listed in `isValidXMLMessage`